src/platform/w3dengine/client/w3ddebugicons.cpp

```diff
@@ -0,0 +1,273 @@
+/**
+ * @file
+ *
+ * @author tomsons26
+ *
+ * @brief Class for drawing small debug icons on screen.
+ * Commonly used for debugging things like pathfinding, trajectories, flight paths.
+ *
+ * @copyright Thyme is free software: you can redistribute it and/or
+ *            modify it under the terms of the GNU General Public License
+ *            as published by the Free Software Foundation, either version
+ *            2 of the License, or (at your option) any later version.
+ *            A full copy of the GNU General Public License can be found in
+ *            LICENSE
+ */
+#include "w3ddebugicons.h"
+#include "gamelogic.h"
+#include "globaldata.h"
+#include "vertmaterial.h"
+#include <new>
+
+#ifdef BUILD_WITH_D3D8
+#include "dx8indexbuffer.h"
+#include "dx8vertexbuffer.h"
+#include "dx8wrapper.h"
+#endif
+
+DebugIcon *W3DDebugIcons::m_debugIcons;
+int W3DDebugIcons::m_numDebugIcons;
+int W3DDebugIcons::m_numUnknown;
+
+W3DDebugIcons::W3DDebugIcons() : m_vertexMaterial(VertexMaterialClass::Get_Preset(VertexMaterialClass::PRELIT_DIFFUSE))
+{
+    Allocate_Icons_Array();
+}
+
+W3DDebugIcons::W3DDebugIcons(const W3DDebugIcons &src)
+{
+    *this = src;
+}
+
+W3DDebugIcons &W3DDebugIcons::operator=(const W3DDebugIcons &that)
+{
+    captainslog_dbgassert(0, "Cannot make a copy of W3DDebugIcons!");
+    return *this;
+}
+
+W3DDebugIcons::~W3DDebugIcons()
+{
+    Ref_Ptr_Release(m_vertexMaterial);
+
+    if (m_debugIcons != nullptr) {
+        delete[] m_debugIcons;
+    }
+
+    m_debugIcons = nullptr;
+    m_numDebugIcons = 0;
+    m_numUnknown = 0;
+}
+
+void W3DDebugIcons::Get_Obj_Space_Bounding_Sphere(SphereClass &sphere) const
+{
+    Vector3 pos(g_theWriteableGlobalData->m_waterExtentX, g_theWriteableGlobalData->m_waterExtentY, 500.0f);
+    sphere.Init(pos, pos.Length());
+}
+
+void W3DDebugIcons::Get_Obj_Space_Bounding_Box(AABoxClass &box) const
+{
+    Vector3 c(-2.0f * g_theWriteableGlobalData->m_waterExtentX, -2.0f * g_theWriteableGlobalData->m_waterExtentY, 0.0f);
+    Vector3 e(2.0f * g_theWriteableGlobalData->m_waterExtentX, 2.0f * g_theWriteableGlobalData->m_waterExtentY, 1000.0f);
+
+    box.Init(c, e);
+}
+
+RenderObjClass *W3DDebugIcons::Clone() const
+{
+    return new W3DDebugIcons;
```

Comment:
This does not look proper. This will call constructor and `Allocate_Icons_Array()` and then overwrite the static members, essentially leaking whatever the previous instance had allocated.